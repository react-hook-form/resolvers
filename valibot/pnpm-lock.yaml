lockfileVersion: '6.0'

settings:
  autoInstallPeers: true
  excludeLinksFromLockfile: false

dependencies:
  '@hookform/resolvers':
    specifier: ^2.0.0
    version: 2.0.0(react-hook-form@7.0.0)
  react-hook-form:
    specifier: ^7.0.0
    version: 7.0.0(react@17.0.2)
  valibot:
    specifier: '>=0.8'
    version: 0.8.0

packages:

  /@hookform/resolvers@2.0.0(react-hook-form@7.0.0):
    resolution: {integrity: sha512-picG6qjP516JNblvg/wuDc8JPfdm3aNbtPbfu+PXQWFXp6ED6dyBL2IoB/tDvPFaNwwqUEwb6UYVKMIFAH+aqQ==}
    peerDependencies:
      react-hook-form: ^7.0.0
    dependencies:
      react-hook-form: 7.0.0(react@17.0.2)
    dev: false

  /js-tokens@4.0.0:
    resolution: {integrity: sha512-RdJUflcE3cUzKiMqQgsCu06FPu9UdIJO0beYbPhHN4k6apgJtifcoCtT9bcxOpYBtpD2kCM6Sbzg4CausW/PKQ==}
    dev: false

  /loose-envify@1.4.0:
    resolution: {integrity: sha512-lyuxPGr/Wfhrlem2CL/UcnUc1zcqKAImBDzukY7Y5F/yQiNdko6+fRLevlw1HgMySw7f611UIY408EtxRSoK3Q==}
    hasBin: true
    dependencies:
      js-tokens: 4.0.0
    dev: false

  /object-assign@4.1.1:
    resolution: {integrity: sha512-rJgTQnkUnH1sFw8yT6VSU3zD3sWmu6sZhIseY8VX+GRu3P6F7Fu+JNDoXfklElbLJSnc3FUQHVe4cU5hj+BcUg==}
    engines: {node: '>=0.10.0'}
    dev: false

  /react-hook-form@7.0.0(react@17.0.2):
    resolution: {integrity: sha512-cXoKnNnQe5IxieesTu+enDS7Jfh8lyX8mSMwzNvC4L0k6R6yQ2F8rIhvBbzlFOGSoeEnsR64dvMZ1izWRCAw6A==}
    peerDependencies:
      react: ^16.8.0 || ^17
    dependencies:
      react: 17.0.2
    dev: false

  /react@17.0.2:
    resolution: {integrity: sha512-gnhPt75i/dq/z3/6q/0asP78D0u592D5L1pd7M8P+dck6Fu/jJeL6iVVK23fptSUZj8Vjf++7wXA8UNclGQcbA==}
    engines: {node: '>=0.10.0'}
    dependencies:
      loose-envify: 1.4.0
      object-assign: 4.1.1
    dev: false

  /valibot@0.8.0:
    resolution: {integrity: sha512-wQHXVkVFj6Z0R3297icGCp3UX8S66onuIg03ihJ8aVgMn8pMJvYSfmrKb+aIg7w/Aw08togrklaMSqDP2vKtIA==}
    dev: false
