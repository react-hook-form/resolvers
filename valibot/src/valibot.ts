import { toNestError } from '@hookform/resolvers';
import type { Resolver } from './types';
import {
  BaseSchema,
  BaseSchemaAsync,
  ValiError,
  parse,
  parseAsync,
} from 'valibot';
import { FieldErrors, FieldError } from 'react-hook-form';

type FlatErrors = Record<string, [FieldError, ...FieldError[]]>;

const parseErrors = (error: ValiError): FieldErrors => {
  const errors = error.issues.reduce<FlatErrors>((flatErrors, issue) => {
    if (issue.path) {
      const path = issue.path.map(({ key }) => key).join('.');
      flatErrors[path] = [
        ...(flatErrors[path] || []),
        {
          message: issue.message,
          type: issue.validation,
        },
      ];
    }

    return flatErrors;
  }, {});

  return Object.entries(errors).reduce<FieldErrors>((acc, [path, errors]) => {
    const [firstError] = errors;
    acc[path] = {
      message: firstError.message,
      type: firstError.type,
    };

    return acc;
  }, {});
};

export const valibotResolver: Resolver =
  (
    schema,
    schemaOptions = {
      abortEarly: false,
      abortPipeEarly: false,
    },
    resolverOptions = {
      mode: 'async',
      raw: false,
    },
  ) =>
  async (values, _, options) => {
    try {
      const { mode, raw } = resolverOptions;
      const parsed =
        mode === 'sync'
          ? parse(schema as BaseSchema, values, schemaOptions)
          : await parseAsync(
              schema as BaseSchema | BaseSchemaAsync,
              values,
              schemaOptions,
            );

      return { values: raw ? values : parsed, errors: {} as FieldErrors };
    } catch (error) {
      if (error instanceof ValiError) {
        return {
          values: {},
          errors: toNestError(parseErrors(error), options),
        };
      }

      throw error;
    }
  };
