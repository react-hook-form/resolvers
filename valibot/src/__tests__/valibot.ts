/* eslint-disable no-console, @typescript-eslint/ban-ts-comment */
import { valibotResolver } from '..';
import { schema, validData, fields, invalidData } from './__fixtures__/data';
import * as valibot from 'valibot';

const shouldUseNativeValidation = false;
describe('valibotResolver', () => {
  it('should return parsed values from valibotResolver with `mode: sync` when validation pass', async () => {
    vi.mock('valibot', async () => {
      const a = await vi.importActual<any>('valibot');
      return {
        __esModule: true,
        ...a,
      };
    });
    const parseSpy = vi.spyOn(valibot, 'parse');
    const parseAsyncSpy = vi.spyOn(valibot, 'parseAsync');

    const result = await valibotResolver(schema, undefined, {
      mode: 'sync',
    })(validData, undefined, { fields, shouldUseNativeValidation });

    expect(parseSpy).toHaveBeenCalledTimes(1);
    expect(parseAsyncSpy).not.toHaveBeenCalled();
    expect(result.errors).toEqual({});
    expect(result).toMatchSnapshot();
  });

  it('should return a single error from valibotResolver with `mode: sync` when validation fails', async () => {
    vi.mock('valibot', async () => {
      const a = await vi.importActual<any>('valibot');
      return {
        __esModule: true,
        ...a,
      };
    });
    const parseSpy = vi.spyOn(valibot, 'parse');
    const parseAsyncSpy = vi.spyOn(valibot, 'parseAsync');

    const result = await valibotResolver(schema, undefined, {
      mode: 'sync',
    })(invalidData, undefined, { fields, shouldUseNativeValidation });

    expect(parseSpy).toHaveBeenCalledTimes(1);
    expect(parseAsyncSpy).not.toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });

  it('should return values from valibotResolver when validation pass', async () => {
    const result = await valibotResolver(schema)(validData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toEqual({ errors: {}, values: validData });
  });

  it('should return a single error from valibotResolver when validation fails', async () => {
    const result = await valibotResolver(schema)(invalidData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toMatchSnapshot();
  });

  it('should return values from valibotResolver when validation pass & raw=true', async () => {
    const result = await valibotResolver(schema, undefined, {
      raw: true,
    })(validData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toEqual({ errors: {}, values: validData });
  });

  it('should return all the errors from valibotResolver when validation fails with `validateAllFieldCriteria` set to true', async () => {
    const result = await valibotResolver(schema)(invalidData, undefined, {
      fields,
      criteriaMode: 'all',
      shouldUseNativeValidation,
    });

    expect(result).toMatchSnapshot();
  });

  it('should return all the errors from valibotResolver when validation fails with `validateAllFieldCriteria` set to true and `mode: sync`', async () => {
    const result = await valibotResolver(schema, undefined, { mode: 'sync' })(
      invalidData,
      undefined,
      {
        fields,
        criteriaMode: 'all',
        shouldUseNativeValidation,
      },
    );

    expect(result).toMatchSnapshot();
  });
});
