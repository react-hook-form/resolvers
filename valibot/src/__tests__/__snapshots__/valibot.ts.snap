// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`valibotResolver > should return a single error from valibotResolver when validation fails 1`] = `
{
  "errors": {
    "accessToken": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
    },
    "birthYear": {
      "message": "Please enter your birth year",
      "ref": undefined,
      "type": "number",
    },
    "email": {
      "message": "Invalid email address",
      "ref": {
        "name": "email",
      },
      "type": "email",
    },
    "enabled": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
    },
    "like": {
      "id": {
        "message": "Like id is required",
        "ref": undefined,
        "type": "number",
      },
      "name": {
        "message": "Invalid type",
        "ref": undefined,
        "type": "non_optional",
      },
    },
    "password": {
      "message": "One uppercase character",
      "ref": {
        "name": "password",
      },
      "type": "regex",
    },
    "repeatPassword": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
    },
    "tags": [
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
      },
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
      },
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
      },
    ],
    "username": {
      "message": "Invalid type",
      "ref": {
        "name": "username",
      },
      "type": "non_optional",
    },
  },
  "values": {},
}
`;

exports[`valibotResolver > should return a single error from valibotResolver with \`mode: sync\` when validation fails 1`] = `
{
  "errors": {
    "accessToken": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
    },
    "birthYear": {
      "message": "Please enter your birth year",
      "ref": undefined,
      "type": "number",
    },
    "email": {
      "message": "Invalid email address",
      "ref": {
        "name": "email",
      },
      "type": "email",
    },
    "enabled": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
    },
    "like": {
      "id": {
        "message": "Like id is required",
        "ref": undefined,
        "type": "number",
      },
      "name": {
        "message": "Invalid type",
        "ref": undefined,
        "type": "non_optional",
      },
    },
    "password": {
      "message": "One uppercase character",
      "ref": {
        "name": "password",
      },
      "type": "regex",
    },
    "repeatPassword": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
    },
    "tags": [
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
      },
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
      },
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
      },
    ],
    "username": {
      "message": "Invalid type",
      "ref": {
        "name": "username",
      },
      "type": "non_optional",
    },
  },
  "values": {},
}
`;

exports[`valibotResolver > should return all the errors from valibotResolver when validation fails with \`validateAllFieldCriteria\` set to true 1`] = `
{
  "errors": {
    "accessToken": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
      "types": {
        "non_optional": "Invalid type",
      },
    },
    "birthYear": {
      "message": "Please enter your birth year",
      "ref": undefined,
      "type": "number",
      "types": {
        "number": "Please enter your birth year",
      },
    },
    "email": {
      "message": "Invalid email address",
      "ref": {
        "name": "email",
      },
      "type": "email",
      "types": {
        "email": "Invalid email address",
      },
    },
    "enabled": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
      "types": {
        "non_optional": "Invalid type",
      },
    },
    "like": {
      "id": {
        "message": "Like id is required",
        "ref": undefined,
        "type": "number",
        "types": {
          "number": "Like id is required",
        },
      },
      "name": {
        "message": "Invalid type",
        "ref": undefined,
        "type": "non_optional",
        "types": {
          "non_optional": "Invalid type",
        },
      },
    },
    "password": {
      "message": "One uppercase character",
      "ref": {
        "name": "password",
      },
      "type": "regex",
      "types": {
        "min_length": "Must be at least 8 characters in length",
        "regex": [
          "One uppercase character",
          "One lowercase character",
          "One number",
        ],
      },
    },
    "repeatPassword": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
      "types": {
        "non_optional": "Invalid type",
      },
    },
    "tags": [
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
        "types": {
          "string": "Tags should be strings",
        },
      },
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
        "types": {
          "string": "Tags should be strings",
        },
      },
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
        "types": {
          "string": "Tags should be strings",
        },
      },
    ],
    "username": {
      "message": "Invalid type",
      "ref": {
        "name": "username",
      },
      "type": "non_optional",
      "types": {
        "non_optional": "Invalid type",
      },
    },
  },
  "values": {},
}
`;

exports[`valibotResolver > should return all the errors from valibotResolver when validation fails with \`validateAllFieldCriteria\` set to true and \`mode: sync\` 1`] = `
{
  "errors": {
    "accessToken": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
      "types": {
        "non_optional": "Invalid type",
      },
    },
    "birthYear": {
      "message": "Please enter your birth year",
      "ref": undefined,
      "type": "number",
      "types": {
        "number": "Please enter your birth year",
      },
    },
    "email": {
      "message": "Invalid email address",
      "ref": {
        "name": "email",
      },
      "type": "email",
      "types": {
        "email": "Invalid email address",
      },
    },
    "enabled": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
      "types": {
        "non_optional": "Invalid type",
      },
    },
    "like": {
      "id": {
        "message": "Like id is required",
        "ref": undefined,
        "type": "number",
        "types": {
          "number": "Like id is required",
        },
      },
      "name": {
        "message": "Invalid type",
        "ref": undefined,
        "type": "non_optional",
        "types": {
          "non_optional": "Invalid type",
        },
      },
    },
    "password": {
      "message": "One uppercase character",
      "ref": {
        "name": "password",
      },
      "type": "regex",
      "types": {
        "min_length": "Must be at least 8 characters in length",
        "regex": [
          "One uppercase character",
          "One lowercase character",
          "One number",
        ],
      },
    },
    "repeatPassword": {
      "message": "Invalid type",
      "ref": undefined,
      "type": "non_optional",
      "types": {
        "non_optional": "Invalid type",
      },
    },
    "tags": [
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
        "types": {
          "string": "Tags should be strings",
        },
      },
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
        "types": {
          "string": "Tags should be strings",
        },
      },
      {
        "message": "Tags should be strings",
        "ref": undefined,
        "type": "string",
        "types": {
          "string": "Tags should be strings",
        },
      },
    ],
    "username": {
      "message": "Invalid type",
      "ref": {
        "name": "username",
      },
      "type": "non_optional",
      "types": {
        "non_optional": "Invalid type",
      },
    },
  },
  "values": {},
}
`;

exports[`valibotResolver > should return parsed values from valibotResolver with \`mode: sync\` when validation pass 1`] = `
{
  "errors": {},
  "values": {
    "accessToken": "accessToken",
    "birthYear": 2000,
    "email": "john@doe.com",
    "enabled": true,
    "like": {
      "id": 1,
      "name": "name",
    },
    "password": "Password123_",
    "repeatPassword": "Password123_",
    "tags": [
      "tag1",
      "tag2",
    ],
    "username": "Doe",
  },
}
`;
